File: crawler/src/fetch.ts
import path from "node:path";

import { File as MegaFile } from "megajs";

import { readJsonFile, resolveInputPath, resolveOutputPath, writeJsonFile } from "./utils/files.ts";
import { dedupeByKey, mapWithConcurrency, normalizeNodeName } from "./utils/index.ts";
import type {
	FetchGroup,
	FetchInputPayload,
	FetchOutputPayload,
	Release,
	ReleaseFile,
} from "./utils/types.ts";

type MegaNode = {
	name: string | null;
	size?: number;
	directory: boolean;
	children?: MegaNode[];
};

type LinkTask = {
	groupIndex: number;
	circle: string;
	link: string;
};

type GroupAccumulator = {
	releaseSets: Release[][];
	errors: string[];
};

const AUDIO_FILE_EXTENSIONS = new Set([
	".aac",
	".aif",
	".aiff",
	".alac",
	".ape",
	".dsf",
	".flac",
	".m4a",
	".mp3",
	".ogg",
	".opus",
	".wav",
	".wma",
]);

const DEFAULT_CONCURRENCY = 4;
const MEGA_LOAD_MAX_ATTEMPTS = 3;
const MEGA_LOAD_RETRY_BASE_DELAY_MS = 750;
const MEGA_LOAD_RETRY_MAX_DELAY_MS = 5000;
const RETRYABLE_MEGA_ERROR_CODES = new Set([
	"ECONNABORTED",
	"ECONNRESET",
	"ENETDOWN",
	"ENETRESET",
	"ENETUNREACH",
	"EAI_AGAIN",
	"ETIMEDOUT",
]);
const RETRYABLE_MEGA_ERROR_MESSAGES = [
	"fetch failed",
	"network",
	"socket",
	"timeout",
	"timed out",
	"temporary failure",
	"connection reset",
];

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => {
		setTimeout(resolve, ms);
	});
}

function formatError(error: unknown): string {
	if (error instanceof Error && error.message.trim().length > 0) {
		return error.message;
	}

	return String(error);
}

function getErrorCode(error: unknown): string | null {
	if (typeof error !== "object" || error === null) return null;

	const value = (error as { code?: unknown }).code;
	if (typeof value !== "string") return null;

	const normalized = value.trim().toUpperCase();
	return normalized.length > 0 ? normalized : null;
}

function isRetryableMegaLoadError(error: unknown): boolean {
	const code = getErrorCode(error);
	if (code && RETRYABLE_MEGA_ERROR_CODES.has(code)) {
		return true;
	}

	const message = formatError(error).toLowerCase();
	return RETRYABLE_MEGA_ERROR_MESSAGES.some((token) => message.includes(token));
}

function getRetryDelayMs(attempt: number): number {
	const exponentialDelay = Math.min(
		MEGA_LOAD_RETRY_MAX_DELAY_MS,
		MEGA_LOAD_RETRY_BASE_DELAY_MS * 2 ** (attempt - 1),
	);
	const jitter = Math.floor(Math.random() * 250);
	return Math.min(MEGA_LOAD_RETRY_MAX_DELAY_MS, exponentialDelay + jitter);
}

function isAudioFileName(name: string): boolean {
	const extension = path.extname(name).toLowerCase();
	return AUDIO_FILE_EXTENSIONS.has(extension);
}

function getDownloadId(node: MegaFile): string {
	return Array.isArray(node.downloadId)
		? node.downloadId[node.downloadId.length - 1]
		: node.downloadId;
}

function collectLeafFiles(node: MegaNode): ReleaseFile[] {
	const name = normalizeNodeName(node.name);

	if (!node.directory && isAudioFileName(name)) {
		const sizeBytes = typeof node.size === "number" ? node.size : 0;
		return [{ name, sizeBytes }];
	}

	if (node.children === undefined) return [];
	return node.children.flatMap((child) => collectLeafFiles(child));
}

function sumFileSizes(files: ReleaseFile[]): number {
	return files.reduce((total, file) => total + file.sizeBytes, 0);
}

async function buildRelease(node: MegaFile, rootLink: string): Promise<Release> {
	const name = normalizeNodeName(node.name);
	const files = node.directory
		? collectLeafFiles(node)
		: [{ name, sizeBytes: typeof node.size === "number" ? node.size : 0 }];

	const sizeBytes = sumFileSizes(files);
	const link = `${rootLink}/folder/${getDownloadId(node)}`;

	return {
		name,
		link,
		directory: node.directory,
		sizeBytes,
		files,
	};
}

async function loadMegaFolder(link: string): Promise<MegaFile> {
	let lastError: unknown = null;

	for (let attempt = 1; attempt <= MEGA_LOAD_MAX_ATTEMPTS; attempt += 1) {
		try {
			const root = MegaFile.fromURL(link);
			const loadedRoot = await root.loadAttributes();
			return loadedRoot as MegaFile;
		} catch (error: unknown) {
			lastError = error;
			const retryable = isRetryableMegaLoadError(error);
			const hasRemainingAttempts = attempt < MEGA_LOAD_MAX_ATTEMPTS;

			if (!retryable || !hasRemainingAttempts) {
				const reason = formatError(error);
				throw new Error(
					`Failed to load MEGA folder after ${attempt} attempt${attempt === 1 ? "" : "s"} (${link}): ${reason}`,
				);
			}

			const nextAttempt = attempt + 1;
			const delayMs = getRetryDelayMs(attempt);
			console.warn(
				`[retry ${nextAttempt}/${MEGA_LOAD_MAX_ATTEMPTS}] Retrying MEGA load in ${delayMs}ms: ${link} (${formatError(error)})`,
			);
			await sleep(delayMs);
		}
	}

	throw new Error(
		`Failed to load MEGA folder after ${MEGA_LOAD_MAX_ATTEMPTS} attempts (${link}): ${formatError(lastError)}`,
	);
}

async function fetchReleasesFromLink(link: string): Promise<Release[]> {
	const root = await loadMegaFolder(link);
	const name = normalizeNodeName(root.name, "Root");

	if (!root.directory) {
		return [await buildRelease(root, link)];
	}

	if (root.children === undefined || root.children.length === 0) {
		return [{ name, link, directory: true, sizeBytes: 0, files: [] }];
	}

	return Promise.all(root.children.map((child) => buildRelease(child, link)));
}

function dedupeReleases(releases: Release[]): Release[] {
	return dedupeByKey(
		releases,
		(release) => `${release.name}::${release.link}::${release.sizeBytes}`,
	);
}

export async function fetchReleases(inputArg?: string, outputArg?: string): Promise<void> {
	const inputPath = resolveInputPath(inputArg, "dist/transformed.json");
	const outputPath = resolveOutputPath(outputArg, "dist/releases.json");
	const inputJson = await readJsonFile<FetchInputPayload>(inputPath);

	if (!Array.isArray(inputJson.groups)) {
		throw new Error("Invalid input JSON: expected top-level 'groups' array.");
	}

	const groups = inputJson.groups;
	let processedLinkCount = 0;
	const totalLinks = groups.reduce((sum, group) => sum + group.links.length, 0);
	const linkTasks: LinkTask[] = groups.flatMap((group, groupIndex) =>
		group.links.map((link) => ({ groupIndex, circle: group.circle, link })),
	);

	const taskResults = await mapWithConcurrency(linkTasks, DEFAULT_CONCURRENCY, async (task) => {
		processedLinkCount += 1;
		console.log(`[${processedLinkCount}/${totalLinks}] Fetching ${task.circle}: ${task.link}`);

		try {
			return {
				groupIndex: task.groupIndex,
				releases: await fetchReleasesFromLink(task.link),
				error: null,
			};
		} catch (error: unknown) {
			const message = error instanceof Error ? error.message : String(error);
			return {
				groupIndex: task.groupIndex,
				releases: [],
				error: `${task.link}: ${message}`,
			};
		}
	});

	const groupAccumulators: GroupAccumulator[] = groups.map(() => ({
		releaseSets: [],
		errors: [],
	}));

	for (const result of taskResults) {
		const accumulator = groupAccumulators[result.groupIndex];
		accumulator.releaseSets.push(result.releases);
		if (result.error) {
			accumulator.errors.push(result.error);
		}
	}

	const outputGroups: FetchGroup[] = groups.map((group, groupIndex) => {
		const accumulator = groupAccumulators[groupIndex];
		const releases = dedupeReleases(accumulator.releaseSets.flat()).sort((a, b) =>
			a.name.localeCompare(b.name),
		);

		return {
			...group,
			releases,
			errors: accumulator.errors,
		};
	});

	const outputJson: FetchOutputPayload = { groups: outputGroups };
	await writeJsonFile(outputPath, outputJson);

	const totalReleases = outputGroups.reduce((sum, group) => sum + group.releases.length, 0);
	console.log(
		`Wrote ${outputGroups.length} groups, ${totalLinks} MEGA links, and ${totalReleases} releases to ${outputPath}`,
	);
}

File: cli/internal/pipeline/fetch/fetch.go
package fetch

import (
	"context"
	"fmt"
	"math/rand"
	"path/filepath"
	"sort"
	"strings"
	"sync/atomic"
	"time"

	"github.com/Curstantine/dcidxr/cli/internal/files"
	"github.com/Curstantine/dcidxr/cli/internal/megapublic"
	"github.com/Curstantine/dcidxr/cli/internal/model"
	"github.com/Curstantine/dcidxr/cli/internal/util"
)

const (
	defaultConcurrency       = 4
	megaLoadMaxAttempts      = 3
	megaLoadRetryBaseDelayMS = 750
	megaLoadRetryMaxDelayMS  = 5000
)

var (
	audioFileExtensions = map[string]struct{}{
		".aac": {}, ".aif": {}, ".aiff": {}, ".alac": {}, ".ape": {}, ".dsf": {},
		".flac": {}, ".m4a": {}, ".mp3": {}, ".ogg": {}, ".opus": {}, ".wav": {}, ".wma": {},
	}
	retryableMegaErrorCodes = map[string]struct{}{
		"ECONNABORTED": {},
		"ECONNRESET":   {},
		"ENETDOWN":     {},
		"ENETRESET":    {},
		"ENETUNREACH":  {},
		"EAI_AGAIN":    {},
		"ETIMEDOUT":    {},
	}
	retryableMegaErrorMessageTokens = []string{
		"fetch failed",
		"network",
		"socket",
		"timeout",
		"timed out",
		"temporary failure",
		"connection reset",
	}
)

type fetchInputPayload struct {
	Groups []model.GroupBase `json:"groups"`
}

type linkTask struct {
	groupIndex int
	circle     string
	link       string
}

type taskResult struct {
	groupIndex int
	releases   []model.Release
	errorText  string
}

type groupAccumulator struct {
	releaseSets [][]model.Release
	errors      []string
}

func Run(ctx context.Context, inputArg, outputArg string) error {
	inputPath := files.ResolvePath(inputArg, "dist/transformed.json")
	outputPath := files.ResolvePath(outputArg, "dist/releases.json")

	inputJSON, err := files.ReadJSONFile[fetchInputPayload](inputPath)
	if err != nil {
		return err
	}

	if inputJSON.Groups == nil {
		return fmt.Errorf("invalid input JSON: expected top-level 'groups' array")
	}

	groups := inputJSON.Groups
	linkTasks := make([]linkTask, 0)
	totalLinks := 0
	for groupIndex, group := range groups {
		totalLinks += len(group.Links)
		for _, link := range group.Links {
			linkTasks = append(linkTasks, linkTask{groupIndex: groupIndex, circle: group.Circle, link: link})
		}
	}

	var processedCount int64
	taskResults, err := util.MapWithConcurrency(ctx, linkTasks, defaultConcurrency, func(ctx context.Context, task linkTask, _ int) (taskResult, error) {
		current := atomic.AddInt64(&processedCount, 1)
		fmt.Printf("[%d/%d] Fetching %s: %s\n", current, totalLinks, task.circle, task.link)

		releases, fetchErr := fetchReleasesFromLink(ctx, task.link)
		if fetchErr != nil {
			return taskResult{
				groupIndex: task.groupIndex,
				releases:   []model.Release{},
				errorText:  fmt.Sprintf("%s: %s", task.link, fetchErr.Error()),
			}, nil
		}

		return taskResult{groupIndex: task.groupIndex, releases: releases}, nil
	})
	if err != nil {
		return err
	}

	groupAccumulators := make([]groupAccumulator, len(groups))
	for index := range groupAccumulators {
		groupAccumulators[index] = groupAccumulator{releaseSets: [][]model.Release{}, errors: []string{}}
	}

	for _, result := range taskResults {
		accumulator := &groupAccumulators[result.groupIndex]
		accumulator.releaseSets = append(accumulator.releaseSets, result.releases)
		if result.errorText != "" {
			accumulator.errors = append(accumulator.errors, result.errorText)
		}
	}

	outputGroups := make([]model.FetchGroup, 0, len(groups))
	for groupIndex, group := range groups {
		accumulator := groupAccumulators[groupIndex]
		flattened := make([]model.Release, 0)
		for _, set := range accumulator.releaseSets {
			flattened = append(flattened, set...)
		}

		releases := dedupeReleases(flattened)
		sort.Slice(releases, func(i, j int) bool {
			return strings.Compare(releases[i].Name, releases[j].Name) < 0
		})

		outputGroups = append(outputGroups, model.FetchGroup{
			GroupBase: group,
			Releases:  releases,
			Errors:    accumulator.errors,
		})
	}

	output := model.FetchOutputPayload{Groups: outputGroups}
	if err := files.WriteJSONFile(outputPath, output); err != nil {
		return err
	}

	totalReleases := 0
	for _, group := range outputGroups {
		totalReleases += len(group.Releases)
	}

	fmt.Printf(
		"Wrote %d groups, %d MEGA links, and %d releases to %s\n",
		len(outputGroups),
		totalLinks,
		totalReleases,
		outputPath,
	)

	return nil
}

func fetchReleasesFromLink(ctx context.Context, link string) ([]model.Release, error) {
	root, err := loadMegaFolder(ctx, link)
	if err != nil {
		return nil, err
	}

	name := util.NormalizeNodeName(root.Name, "Root")
	if !root.Directory {
		release := buildRelease(root, link)
		if release.Name == "" {
			release.Name = name
		}
		return []model.Release{release}, nil
	}

	if len(root.Children) == 0 {
		return []model.Release{{Name: name, Link: link, Directory: true, SizeBytes: 0, Files: []model.ReleaseFile{}}}, nil
	}

	releases := make([]model.Release, 0, len(root.Children))
	for _, child := range root.Children {
		releases = append(releases, buildRelease(child, link))
	}

	return releases, nil
}

func loadMegaFolder(ctx context.Context, link string) (*megapublic.Node, error) {
	var lastErr error

	for attempt := 1; attempt <= megaLoadMaxAttempts; attempt++ {
		node, err := megapublic.Load(ctx, link)
		if err == nil {
			return node, nil
		}

		lastErr = err
		retryable := isRetryableMegaLoadError(err)
		hasRemainingAttempts := attempt < megaLoadMaxAttempts
		if !retryable || !hasRemainingAttempts {
			return nil, fmt.Errorf(
				"failed to load MEGA folder after %d attempt%s (%s): %s",
				attempt,
				pluralize(attempt),
				link,
				err.Error(),
			)
		}

		nextAttempt := attempt + 1
		delay := retryDelayMS(attempt)
		fmt.Printf(
			"[retry %d/%d] Retrying MEGA load in %dms: %s (%s)\n",
			nextAttempt,
			megaLoadMaxAttempts,
			delay,
			link,
			err.Error(),
		)

		timer := time.NewTimer(time.Duration(delay) * time.Millisecond)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, ctx.Err()
		case <-timer.C:
		}
	}

	return nil, fmt.Errorf(
		"failed to load MEGA folder after %d attempts (%s): %s",
		megaLoadMaxAttempts,
		link,
		lastErr.Error(),
	)
}

func buildRelease(node *megapublic.Node, rootLink string) model.Release {
	name := util.NormalizeNodeName(node.Name, "Unknown")

	var filesOut []model.ReleaseFile
	if node.Directory {
		filesOut = collectLeafFiles(node)
	} else {
		filesOut = []model.ReleaseFile{{Name: name, SizeBytes: node.Size}}
	}

	size := int64(0)
	for _, file := range filesOut {
		size += file.SizeBytes
	}

	return model.Release{
		Name:      name,
		Link:      fmt.Sprintf("%s/folder/%s", rootLink, node.DownloadID),
		Directory: node.Directory,
		SizeBytes: size,
		Files:     filesOut,
	}
}

func collectLeafFiles(node *megapublic.Node) []model.ReleaseFile {
	name := util.NormalizeNodeName(node.Name, "Unknown")
	if !node.Directory && isAudioFileName(name) {
		return []model.ReleaseFile{{Name: name, SizeBytes: node.Size}}
	}

	if len(node.Children) == 0 {
		return []model.ReleaseFile{}
	}

	files := make([]model.ReleaseFile, 0)
	for _, child := range node.Children {
		files = append(files, collectLeafFiles(child)...)
	}

	return files
}

func dedupeReleases(values []model.Release) []model.Release {
	return util.DedupeByKey(values, func(value model.Release) string {
		return fmt.Sprintf("%s::%s::%d", value.Name, value.Link, value.SizeBytes)
	})
}

func isAudioFileName(name string) bool {
	ext := strings.ToLower(filepath.Ext(name))
	_, ok := audioFileExtensions[ext]
	return ok
}

func isRetryableMegaLoadError(err error) bool {
	message := strings.ToLower(err.Error())
	for code := range retryableMegaErrorCodes {
		if strings.Contains(message, strings.ToLower(code)) {
			return true
		}
	}

	for _, token := range retryableMegaErrorMessageTokens {
		if strings.Contains(message, token) {
			return true
		}
	}

	return false
}

func retryDelayMS(attempt int) int {
	delay := megaLoadRetryBaseDelayMS * (1 << max(0, attempt-1))
	if delay > megaLoadRetryMaxDelayMS {
		delay = megaLoadRetryMaxDelayMS
	}

	jitter := rand.Intn(250)
	delay += jitter
	if delay > megaLoadRetryMaxDelayMS {
		delay = megaLoadRetryMaxDelayMS
	}

	return delay
}

func pluralize(value int) string {
	if value == 1 {
		return ""
	}

	return "s"
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}
